DocumentDB guide (#43888)
gravitational/teleport: docs/pages/enroll-resources/database-access/enroll-aws-databases/aws-docdb.mdx

<Tabs>
<TabItem label="Self-Hosted">
![Teleport Architecture DocumentDB Access Self-Hosted](../../../../img/database-access/guides/docdb_selfhosted.svg)
</TabItem>
<TabItem label="Cloud-Hosted">
![Teleport Architecture DocumentDB Cloud](../../../../img/database-access/guides/docdb_cloud.svg)
</TabItem>

</Tabs>

(!docs/pages/includes/database-access/auto-discovery-tip.mdx dbType="Amazon DocumentDB cluster" providerType="AWS"!)

## Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

- AWS account with a DocumentDB cluster.
  <Admonition type="warning" title="IAM authentication and TLS">
  DocumentDB access requires IAM authentication which is only supported for
  DocumentDB 5.0 or higher.

  In addition, the cluster must have Transport Layer Security (TLS) enabled.
  TLS is enabled by default when using the default parameter group.
  </Admonition>
- Command-line client `mongosh` or `mongo` installed and added to your
  system's `PATH` environment variable.
- A host, e.g., an EC2 instance, where you will run the Teleport Database
  Service.
- (!docs/pages/includes/tctl.mdx!)

## Step 1/6. Create a Teleport user

> (!docs/pages/includes/database-access/create-user.mdx!)

Author: yes, it's at the protocol level.

> <Admonition type="warning">                                
>  Database names are only enforced for PostgreSQL, MongoDB, and Cloud Spanner databases.                                                   
> </Admonition> 

Are you thinking of updating this?

> (!docs/pages/includes/database-access/create-user.mdx!)

Reviewer: yeah updating that to include docdb

> (!docs/pages/includes/database-access/create-user.mdx!)

Author: But we don't include like RDS there?

> (!docs/pages/includes/database-access/create-user.mdx!)

Reviewer: oh, I guess I keep forgetting that docdb is mongo - no need to update that then

        mechanisms: ["MONGODB-AWS"],
        roles: [ { role: "root", db: "admin" },  ]
    }
);
```

This example creates a DocumentDB IAM user with root privileges. You can adjust
the privileges as required.

<Admonition type="tip" title="Avoid password authentication">
This step is a one-time setup that must use password authentication. However,
after creating the first IAM user, you can skip password authentication by
accessing DocumentDB through Teleport and creating more DocumentDB IAM users
using the first IAM user.
</Admonition>

## Step 6/6 Connect

Once the Database Service has started and joined the cluster, log in to see the
registered databases:

```code
$ tsh login --proxy=<Var name="example.teleport.sh:443" /> --user=alice
$ tsh db ls
# Name     Description Labels
# -------- ----------- --------
# my-docdb             env=dev
```

Retrieve credentials for the database and connect to it as the `teleport-docdb-user` user:

```code
$ tsh db connect --db-user=teleport-docdb-user --db-name=test my-docdb
```

Log out of the database and remove credentials:

```code
$ tsh db logout rds-example
```

## Troubleshooting

> (!docs/pages/includes/database-access/aws-troubleshooting.mdx!)

Reviewer: This isn't the fault of this PR, but we should really split out or template the db specific troubleshooting.
I was surprised to see certificate error troubleshooting for RDS in the rendered preview.

Does DocumentDB have this issue? I assume there's some CA bundle for it, but I've never looked at DocumentDB before.

Author: Good catch. DocumentDB uses the same bundle as RDS. I will update the error description there to include DocumentDB.

### Unable to authenticate using mechanism "MONGODB-AWS" error

You may encounter the following error when IAM authentication fails:
```code
$ tsh db connect --db-user=teleport-docdb-user --db-name=test my-docdb
...
MongoServerSelectionError: error connecting to the database
	connection() error occurred during connection handshake: auth error: sasl conversation error: unable to authenticate using mechanism "MONGODB-AWS":  Authentication failed.
ERROR: exit status 1
```

Make sure the IAM role exists as a DocumentDB user in the "$external" database
of the target DocumentDB cluster and the DocumentDB user is configured with
`MONGODB-AWS` mechanisms for authentication.

See [Authentication using IAM
identity](https://docs.aws.amazon.com/documentdb/latest/developerguide/iam-identity-auth.html)
for more details.

## Next steps

> (!docs/pages/includes/database-access/guides-next-steps.mdx!)
> - Learn more about [Authentication using IAM identity for

Reviewer: generally, includes should always have whitespace before and after them.
It seems to be fine here, but sometimes it causes rendering issues.

```suggestion
(!docs/pages/includes/database-access/guides-next-steps.mdx!)

- Learn more about [Authentication using IAM identity for
```

The Database Service will proxy the Amazon DocumentDB cluster with the URI you
specified earlier. Keep in mind that AWS IAM changes may not propagate
immediately and can take a few minutes to come into effect.

## Step 5/6. Create an IAM role for a DocumentDB user

In this step, we will create an IAM role that the Teleport Database Service can
assume to authenticate to DocumentDB on behalf of a Teleport user.

### Create an IAM role

Navigate back to the Roles page on the AWS Web Console and create a new role.
Select the "AWS account" option, which creates a default trust policy to allow
other entities in this account to assume this role:

![Create Role Step 1](../../../../img/database-access/guides/dynamodb-create-role-1.png)

Skip the "Add Permissions" page by clicking "Next", and give the role a name.
In this guide, we will use the example "teleport-docdb-user" for this role.

Now click **Add new tag** at Step 3, use `TeleportDatabaseService` for the key
and `Allowed` for the value. Then click **Create Role** to complete the process.

![Create Role Step 3](../../../../img/database-access/guides/aws-create-role-add-tags.png)

### Create a DocumentDB user

> Login your DocumentDB cluster with your master username and password from a

Reviewer 2: ```suggestion
Log in to your DocumentDB cluster with your master username and password from a
```
Grammar suggestion

machine that has network access to the DocumentDB cluster. Create a DocumentDB
user with the IAM role ARN as username and specify `MONGODB-AWS` in the

> mechanisms for authentication:

Reviewer 2: ```suggestion
mechanisms for authentication:

```

To make the source slightly easier to read

    }
);
```

This example creates a DocumentDB IAM user with root privileges. You can adjust
the privileges as required.

<Admonition type="tip" title="Avoid password authentication">
This step is a one-time setup that must use password authentication. However,
after creating the first IAM user, you can skip password authentication by
accessing DocumentDB through Teleport and creating more DocumentDB IAM users
using the first IAM user.
</Admonition>

## Step 6/6 Connect

Once the Database Service has started and joined the cluster, log in to see the
registered databases:

```code
$ tsh login --proxy=<Var name="example.teleport.sh:443" /> --user=alice
$ tsh db ls
# Name     Description Labels
# -------- ----------- --------
# my-docdb             env=dev
```

Retrieve credentials for the database and connect to it as the `teleport-docdb-user` user:

```code
$ tsh db connect --db-user=teleport-docdb-user --db-name=test my-docdb
```

Log out of the database and remove credentials:

```code
$ tsh db logout rds-example
```

## Troubleshooting

(!docs/pages/includes/database-access/aws-troubleshooting.mdx!)

### Unable to authenticate using mechanism "MONGODB-AWS" error

> You may encounter the following error when IAM authentication fails:

Reviewer 2: ```suggestion
You may encounter the following error when IAM authentication fails:

```
